Update vtk to 9.5

biagas opened the pull request:
### Description

This PR makes VTK version 9.5.0 the supported version.  Support for older VTK has been removed.

Regression test baselines will need to be assessed after the merge and the nightly run.

Misc changes:
- poodle, rzhound and windows config-site files were updated to support the new library versions.
- visit_vtk/offscreen has been removed since VTK determines offscreen context at runtime.
- vtkVisItDataSetMapper has been removed, the necessary functionality being used for cinema support was moved to VisWinRendering.
- vtkVisItStructuredGrid, vtkVisItRectilineaGrid have been removed.
- CMake/VisItObjectFactory* files were removed, they weren't being used.
- OpenEXR has been disabled temporarily until version can be updated (See #20476)
- A couple of blueprint_export regressions were rebaselined.

build_visit changes:

- The check_opengl_context function was updated to first check libOpenGL, then libGL.
- OSMesa and LLVM were added to the required build_visit modules, so that we provide a runtime fallback for offscreen rendering if VTK cannot otherwise generate a good context.
- Logic was added to disable OSMESA if MESGAL is enabled since it will be provided in the MESAGL build.
- Logic was also added to disable LLVM if both OSMESA and MESAGL were turned off.
- ANARI was updated to 0.14.1
- ZLIB was updated to 1.3.1
- CMAKE was updated to 3.31.8


Resolves #20305
Resolves #19816 
Resolves #19857
Resolves #20463
Resolves #20454


### Type of change

* ~~[ ] Bug fix~~
* ~~[ ] New feature~~
* ~~[ ] Documentation update~~
* [X] Other
VTK library version update.

### How Has This Been Tested?

Compiled and ran regressions on poodle and windows. No crashes, though rebaselines will be needed.


### Checklist:

- [X] I have commented my code where applicable.~~
- [X] I have updated the release notes.~~
- ~~[ ] I have made corresponding changes to the documentation.~~
- ~~ [ ] I have added debugging support to my changes.~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works.~~
- ~~[ ] I have confirmed new and existing unit tests pass locally with my changes.~~
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- ~~[ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

biagas committed:
Add bv_vtk95.sh optional module to build 9.5.0 (currently .rc1).
Add --vtk95 to build_visit command line to build.

biagas committed:
Disable OpenEXR with VTK 9.5, as its headers aren't compatible with VTK-9.5's forced linking with c++17 standard.

biagas committed:
Change how dimensions are retrieved from vtkStructuredGrid.
VTK-9.5 removed the `int * vtkStructuredGrid::GetDimensions()` function.
Need to pass in the sized array to `void vtkStructuredGrid::GetDimenstions(int[3])` instead.

biagas committed:
When using vtk 9.5, change use of vtkPolygon::Triangulate to vtkPolygon::TriangulateLocalIds.

biagas committed:
Use GetLinks instead of GetCellLinks with vtk 9.5.

biagas committed:
Use `.c_str()` to access the char* rep of vtkStdString (MFIX reader).

biagas committed:
Set up Xdmf to use vtk-9.5.

biagas committed:
Modify poodle config-site for VTK 9.5.

cyrush committed:
adj build visit for macos and c++17 fixes

cyrush committed:
Merge remote-tracking branch 'origin/biagas/update_vtk_to_9.5' into biagas/update_vtk_to_9.5

cyrush committed:
vtk 95 build fix for macos

biagas committed:
Fix compile error on Windows due to c++17.

Setting WIN32_LEAN_AND_MEAN prevents headers that may not be needed from being included in the build.

griffin28 committed:
ANARI Dynamic UI

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Make vtk95 the required library.
Remove qt from required list since we really only want qt6.

griffin28 committed:
Merge branch 'biagas/update_vtk_to_9.5' of github.com:visit-dav/visit into griffin28/biagas_vtk_to_9.5

griffin28 committed:
Made indent spacing consistent for RegisterOverride calls

cyrush committed:
fix for python filter env

cyrush committed:
Merge remote-tracking branch 'origin/biagas/update_vtk_to_9.5' into biagas/update_vtk_to_9.5

griffin28 committed:
Merge pull request #20419 from visit-dav/griffin28/biagas_vtk_to_9.5

ANARI Dynamic UI

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Update VTK 95 version to rc3.

biagas committed:
Remove visit_vtk/offscreen, no longer needed.

biagas committed:
Fix typos.

biagas committed:
Remove support for VTK versions < 9.5.0.

biagas committed:
Fix typo in vtkDataWriter patch.

biagas committed:
Fix logic that allowed anari to clobber ospray setting of its pass.

biagas committed:
Remove vtkVisItDataSetMapper.
It was only really used for cinema, so moved necessary logic to
VisWinRendering::ScreenRender.
Update places where it was used in RegisterOverride.
Update vtk version to 9.5.0 (final release version).

biagas committed:
Remove print statement that printed to terminal in unv reader.

cyrush committed:
updates for macOS 15.5

cyrush committed:
complete merge

biagas committed:
Fix bv_qt6_enable.

biagas committed:
VTK has a new group designation, disable it along with the rest.

biagas committed:
Remove old vtk build module.

biagas committed:
Rename bv_vtk95.sh to bv_vtk.sh.
Update modules and bv_xdmf accordingly.

biagas committed:
Move bv_qt6.sh to bv_qt.sh.
Update bv_qwt and modules accordingly.

biagas committed:
Update bv_vtk for change of bv_qt6 to bv_qt.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Fix a couple of qt6 references.

biagas committed:
Fix how OSMesa is enabled when VTK is built, but not MesaGL.
Update release notes.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Update rzwhippet10.cmake for vtk 9.5, visit 3.5.0.

biagas committed:
Update poodle config-site for vtk 9.5 visit 3.5.0.

biagas committed:
Update windows config-site file for VTK 9.5.

Also update conduit to v0.9.4, mfem to 4.8.
Simplified some login in FindVTK9.

biagas committed:
Remove unnecessary includes of vtkObjectFactory header.

biagas committed:
Moved some vtkObjectFactory logic in New methods to vtkStandardNewMacro.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Update ci container with latest versions of cmake, vtk, zlib, conduit, mfem.

biagas committed:
Update blueprint_export baselines to see if CI will succeed.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

griffin28 committed:
Resolves #20455

Fixed crash when ANARI rendering selected, but valid back-end found

griffin28 committed:
Merge pull request #20462 from visit-dav/griffin28/biagas/backend

Fixed crash when ANARI rendering selected, but valid back-end found

griffin28 committed:
Removed test code and put in correct logic for setting library name in the UI

griffin28 committed:
Merge pull request #20464 from visit-dav/griffin28/biagas/bv_anari

Removed test code and put in correct logic for setting library name

biagas committed:
Update container tag.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Fix typo.

griffin28 committed:
Resolves #20463

Updated bv_anari to use ANARI v0.14.1 to support the ignoreAmbientLighting flag

griffin28 committed:
Merge pull request #20467 from visit-dav/griffin28/biagas/bv_anari

Updated bv_anari to use ANARI v0.14.1 to support the ignoreAmbientLighting flag

biagas committed:
Perform opengl_context_check against libOpenGL if available.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Update anari, openexr, vtkm in  poodle config-site.

biagas committed:
Update windows config vtkm to 2.3.0

Update ci container tag.

biagas committed:
Make OSMesa,LLVM be required.
OSMesa is needed as a runtime fallback for offscreen rendering if
VTK cannot generated a good context.
However, if MESAGL is needed for onscreen rendering, then disable OSMESA,
since it will be provided in the MESAGL build.
If neither OSMESA nor MESAGL are buing built, turn off LLVM.

biagas committed:
VTK doesn't technically depend on OSMESA.

biagas committed:
Remove no-longer used files.

biagas committed:
Remove support for older vtk.

biagas committed:
Remove references to deleted files.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

markcmiller86 commented on src/CMake/SetUpThirdParty.cmake:
```diff
@@ -626,7 +626,12 @@ if(NOT VISIT_BUILD_MINIMAL_PLUGINS OR VISIT_SELECTED_DATABASE_PLUGINS)
 
-    include(${VISIT_SOURCE_DIR}/CMake/FindOpenEXR.cmake)
+    if(VTK_VERSION VERSION_LESS "9.5.0")
+        # linking with vtk 9.5 imposes c++17, and there
```
Do we really need this logic? We're at VTK 9.5. after this PR. There is no going back is there?

biagas replied:
It was how I commented out OpenEXR. It will be changed once the PR for newer openexr is  merged. Yes, I could have done it a different way.

JustinPrivitera commented on src/avt/Database/Database/avtGenericDatabase.C:
```diff
@@ -5363,7 +5363,8 @@ avtGenericDatabase::MaterialSelect(vtkDataSet *ds, avtMaterial *mat,
     if (ds->GetDataObjectType() == VTK_STRUCTURED_GRID)
     {
-        int *dims = ((vtkStructuredGrid*)ds)->GetDimensions();
+        int dims[3];
+        ((vtkStructuredGrid*)ds)->GetDimensions(dims);
```
why this change?

markcmiller86 commented on src/avt/VisWindow/Colleagues/VisWinRendering.C:
```diff
@@ -315,22 +323,18 @@ VisWinRendering::VisWinRendering(VisWindowColleagueProxy &p) :
-#if defined(HAVE_OSPRAY)
     osprayRendering = false;
+    viewIs3D = true;
+#if defined(HAVE_OSPRAY)
```
Any reason we're using `#if defined(HAVE_XXX)` here but `#ifdef HAVE_XXX` below?
```suggestion
#ifdef HAVE_OSPRAY
```

JustinPrivitera commented on src/engine/main/Engine.C:
```diff
@@ -807,15 +803,7 @@ Engine::InitializeCompute()
-        debug1 << mName << "Offscreen rendering will use GL." << endl;
-#endif
-#endif
+        // any message?
```
any message?

biagas replied:
Thanks, that was a note to remind me to figure out if there should be a message here. I cannot think of one because VTK now determines the backend for offscreen rendering, which I don't think I can query at this point.

biagas replied:
Now I'm wondering if the `SoftwareRendering` flag is still needed.  Will investigate.

biagas replied:
Added a message. Thanks again for pointing this out.

biagas committed:
Add message to Engine about setting up for offscreen rendering.

Even if we can't determine which backend will be used at this point.

markcmiller86 commented on src/avt/VisWindow/Colleagues/VisWinRendering.h:
```diff
@@ -495,38 +470,24 @@ class VISWINDOW_API VisWinRendering : public VisWinColleague
+
     bool                          osprayRendering {false};
+    bool                          viewIs3D {true};
+#if defined(HAVE_OSPRAY)
```
```suggestion
#ifdef HAVE_OSPRAY
```

biagas replied:
I only moved the 'osprayRendering' flag outside of the if. You want me to change what was already there?

cyrush committed:
dont compile vs staggered field, it has missing methods and fails on macos even though templates arent used

cyrush committed:
Merge remote-tracking branch 'origin/biagas/update_vtk_to_9.5' into biagas/update_vtk_to_9.5

markcmiller86 commented on src/operators/IntegralCurve/avtIntegralCurveFilter.C:
```diff
@@ -1696,6 +1699,7 @@ avtIntegralCurveFilter::GenerateSeedPointsFromPlane(std::vector<avtVector> &pts)
 
     if (randomSamples)
     {
+        std::mt19937 rng(std::time(nullptr));
```
Do we care about reproducibility here at all? It looks to me like we're initializing the RNG with the current time.

biagas replied:
@cyrush could you address?

cyrush commented:
Pushed fix for visschema plugin, this will also resolve #20454

markcmiller86 commented on src/test/tests/unit/compiler_warnings_skips.json:
Oh dear. This file is obsolete I think. We should remove it. I am sorry it wound up taking any of your attention.

markcmiller86 commented on src/tools/dev/scripts/bv_support/bv_vtk.sh:
Reviewing this file, I am reminded of some stuff @JustinPrivitera had to do to *default* behavior for VTK file format reading and I just wanna make sure that is also handled (in whatever way necessary) here on VTK-9.5

biagas replied:
Are you talking about the VTK file writing, where we prefer to keep it at the older version? That is handled by an override of vtkDataSetWriter.

markcmiller86 replied:
Yep. That is what I was thinking of. Ok, good!

JustinPrivitera replied:
I think there was a patch I needed to apply? I don't remember. New versions of VTK wouldn't require the patch IIRC.

biagas replied:
There was. When I added support for VTK -9.4, I added an override for vtkDataSetWriter that sets the default version (because VTK's default version is the newer). From PR #20199.

https://github.com/visit-dav/visit/blob/991710863ffda2d3043cd21a5d6814f0d88bc083/src/visit_vtk/full/vtkVisItDataSetWriter.C#L11-L14

JustinPrivitera replied:
Excellent.

markcmiller86 reviewed (commented):
I just had a few comments before approving.

cyrush commented on src/tools/dev/scripts/bv_support/bv_vtk.sh:
```diff
@@ -63,27 +27,23 @@ function bv_vtk_depends_on
-    elif [[ "$DO_OSMESA" == "yes" ]]; then
-        depends_on="${depends_on} osmesa"
+    #elif [[ "$DO_OSMESA" == "yes" ]]; then
+    #    depends_on="${depends_on} osmesa"
```
should we remove this case?

biagas replied:
We have logic for building VTK only against OSMesa. Not sure if there is still a use-case for that, so I left this in (and the OSMESA only logic elsewshere), in case we needed to re-enable that.
I can be convinced to remove it all.

cyrush reviewed (dismissed):
Looks good! One question about a commented out else if in bv_vtk.sh

biagas committed:
Ensure plugin-vs-install tests work.

JustinPrivitera commented on src/tools/dev/diagnostics/CMakeLists.txt:
```diff
@@ -15,8 +15,7 @@ add_subdirectory(networktest)
 
-# OSMesa should always be available with VTK 9.4.1
-if(NOT WIN32 AND (HAVE_OSMESA OR VTK_VERSION VERSION_GREATER_EQUAL "9.4.1"))
+if(LINUX AND HAVE_OSMESA)
```
is this meant to be for mac too?

biagas replied:
Mac doesn't uses OSMesa as far as I recall.  Perhaps just the test for HAVE_OSMESA is sufficient.

biagas replied:
Modified if test to if(OSMESA_FOUND), which should only be true on Linux.

JustinPrivitera commented on test/baseline/databases/blueprint_export/ucd2d.silo.csv-vertex_data.csv.txt:
```diff
@@ -1,10 +1,10 @@
+2.5, 1.5, 2.5, 1.5, 0, 0, 1
+2, 1.5, 2, 1.5, 0, 0, 2
 1.5, 2, 1.5, 2, 0, 0, 3
-2.5, 1.5, 2.5, 1.5, 0, 0, 4
+2, 2, 2, 2, 0, 0, 4
 2.5, 2, 2.5, 2, 0, 0, 5
-2, 2.5, 2, 2.5, 0, 0, 6
-1.5, 2.5, 1.5, 2.5, 0, 0, 7
+1.5, 2.5, 1.5, 2.5, 0, 0, 6
+2, 2.5, 2, 2.5, 0, 0, 7
```
how come the order of these things changed?

biagas replied:
Internal changes in VTK.

JustinPrivitera replied:
I know they changed how unstructured grids are represented internally. My memory is foggy on this but I thought we ran into this with the VTK export legacy vs modern format, and we wanted legacy because things were changed.

I'm just confused that for blueprint export things are looking different, but it looks like the same data, just in a different order.

biagas replied:
This isn't related to the vtk version  used by the vtkDataSetWriter.  It is something fundamentally different about node ordering for unstructured grids. I didn't dive into code to find out exactly why, so I don't know if it is internal to the dataset or if it is a filter somewhere. I am guessing a filter. 
 A lot of pick tests will need to rebaselined due to this, as well as the blueprint export tests I did here. I wanted the blueprint export rebaselines in this PR so that you and @cyrush could see and comment.

JustinPrivitera replied:
Right. Like I said, my memory is foggy. This seems fine to me; maybe @cyrush has more useful feedback to give.

cyrush replied:
The other export changes impacted the style of vtk ascii files. 

These baselines are for flattened data - so we are ok. 

There was an order change in vtk implementation.

JustinPrivitera commented on test/baseline/databases/blueprint_export/unstructured_multi_rect3d.silo.csv-vertex_data.csv.txt:
```diff
@@ -1,52 +1,52 @@
+0.5, 0.333333, 0, 0.5, 0.333333, 0, 0, 6
+0.75, 0.333333, 0, 0.75, 0.333333, 0, 0, 7
+1, 0.333333, 0, 1, 0.333333, 0, 0, 8
+0, 0.666667, 0, 0, 0.666667, 0, 0, 9
```
same question here

JustinPrivitera reviewed (dismissed):
Incredible work.

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Merge branch 'develop' into biagas/update_vtk_to_9.5

biagas committed:
Update support for OpenEXR 3.3.4.
Update poodle and rzwhippet config-site files to include SETUP_APP_VERSION for OpenEXR.
Update FilterDependencies support for OpenEXR.
Remove support for older OpenEXR from its find module.

biagas committed:
Cleanup rzwhippet10.cmake.
Fix test for adding osmesa diagnostic test to build.

biagas committed:
Update windows config site for anari 0.14.1.

biagas commented:
did some minor cleanup which dismissed prior reviews. anyone want to double-check the recent changes before I merge?
